In [1]:
import pandas as pd
from datetime import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import pickle
import re
%matplotlib inline

In [2]:
df = pd.read_csv('USvideos.csv')

In [3]:
df.head()

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  puqaWrEC7tY      17.14.11   
4  d380meD0W0M      17.14.11   

                                               title          channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE           CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...        LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...           Rudy Mancuso   
3                   Nickelback Lyrics: Real or Fake?  Good Mythical Morning   
4                           I Dare You: GOING BALD!?               nigahiga   

   category_id              publish_time  \
0           22  2017-11-13T17:13:01.000Z   
1           24  2017-11-13T07:30:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-13T11:00:04.000Z   
4           24  2017-11-12T18:01:41.000Z   

                                                tags    views   likes  \
0                                    SHANtell martin   748374   57527   
1  last week tonight trump presidency|"last week ...  2418783   97185   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434  146033   
3  rhett and link|"gmm"|"good mythical morning"|"...   343168   10172   
4  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...  2095731  132235   

   dislikes  comment_count                                  thumbnail_link  \
0      2966          15954  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1      6146          12703  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3       666           2146  https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg   
4      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...  
1  One year after the presidential election, John...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  Today we find out if Link is a Nickelback amat...  
4  I know it's been a while since we did this sho...

We see that our dataset contains many columns: some to identify the video and channel that published the video, some to keep track of viewer statistics, and some to describe the kind of video it is. The dataset also contains a 'thumbnail_link' column which contains a link to a page that contains the thumbnail for the video. Although it would be an interesting assignment to examine the picture and use it to understand how the choice of thumbnails can affect the likelihood of trending of a video, I think that might be too complicated an endeavor for me to currently embark on and so will remove it from the dataset.

In [4]:
cats_US = pd.read_json('UScategoryids.json')

In [5]:
cats_US.head()

kind  \
0  youtube#videoCategoryListResponse   
1  youtube#videoCategoryListResponse   
2  youtube#videoCategoryListResponse   
3  youtube#videoCategoryListResponse   
4  youtube#videoCategoryListResponse   

                                                etag  \
0  "m2yskBQFythfE4irbTIeOgYYfBU/S730Ilt-Fi-emsQJv...   
1  "m2yskBQFythfE4irbTIeOgYYfBU/S730Ilt-Fi-emsQJv...   
2  "m2yskBQFythfE4irbTIeOgYYfBU/S730Ilt-Fi-emsQJv...   
3  "m2yskBQFythfE4irbTIeOgYYfBU/S730Ilt-Fi-emsQJv...   
4  "m2yskBQFythfE4irbTIeOgYYfBU/S730Ilt-Fi-emsQJv...   

                                               items  
0  {'kind': 'youtube#videoCategory', 'etag': '"m2...  
1  {'kind': 'youtube#videoCategory', 'etag': '"m2...  
2  {'kind': 'youtube#videoCategory', 'etag': '"m2...  
3  {'kind': 'youtube#videoCategory', 'etag': '"m2...  
4  {'kind': 'youtube#videoCategory', 'etag': '"m2...

We see that cats['items'] appears to be a dictionary with a lot of information.

In [6]:
cats_US['items'][0]

{'kind': 'youtube#videoCategory',
 'etag': '"m2yskBQFythfE4irbTIeOgYYfBU/Xy1mB4_yLrHy_BmKmPBggty2mZQ"',
 'id': '1',
 'snippet': {'channelId': 'UCBR8-60-B28hp2BmDPdntcQ',
  'title': 'Film & Animation',
  'assignable': True}}

Our assumption was correct! cats['items'] is a dictionary that contains various useful bits of information, including the category id and the category title, which, interestingly enough, is located in 'snippet' (which is a dictionary within a dictionary within a json).

In [7]:
categories_US = {int(category['id']): category['snippet']['title'] for category in cats_US['items']}

In [8]:
#Converting from objects to datetime64

df.trending_date = pd.to_datetime(df.trending_date, format='%y.%d.%m', errors='coerce')
df.publish_time = pd.to_datetime(df.publish_time, format='%Y-%m-%dT%H:%M:%S.%fZ', errors='coerce')

In [9]:
#Mapping values in a dictionary to its corresponding keys in the dataframe

df['category_name'] = df['category_id'].map(categories_US)

In [10]:
len(df.category_id.unique())

16

In [11]:
#Reordering columns, replacing 'category_id' with 'category_name', and removing 'thumbnail_link'

df = df[['video_id', 'trending_date', 'publish_time', 'channel_title', 'category_id', 'category_name', 'title', 'description', 'tags', 'views', 'likes', 'dislikes', 'comment_count']]

In [12]:
print("Our American data, on the other hand, covers videos that trended at any point from {} to {}, and the videos could have been published at any point between {} and {}.".format(min(df.trending_date.dt.date), max(df.trending_date.dt.date), min(df.publish_time.dt.date), max(df.publish_time.dt.date)))

Our American data, on the other hand, covers videos that trended at any point from 2017-11-14 to 2018-06-14, and the videos could have been published at any point between 2006-07-23 and 2018-06-14.


In [13]:
df.shape

(40949, 13)

We see that there are 40949 rows and 17 columns in df_US, and there are 40881 rows and 17 columns in df_CA.

In [14]:
df.describe()

category_id         views         likes      dislikes  comment_count
count  40949.000000  4.094900e+04  4.094900e+04  4.094900e+04   4.094900e+04
mean      19.972429  2.360785e+06  7.426670e+04  3.711401e+03   8.446804e+03
std        7.568327  7.394114e+06  2.288853e+05  2.902971e+04   3.743049e+04
min        1.000000  5.490000e+02  0.000000e+00  0.000000e+00   0.000000e+00
25%       17.000000  2.423290e+05  5.424000e+03  2.020000e+02   6.140000e+02
50%       24.000000  6.818610e+05  1.809100e+04  6.310000e+02   1.856000e+03
75%       25.000000  1.823157e+06  5.541700e+04  1.938000e+03   5.755000e+03
max       43.000000  2.252119e+08  5.613827e+06  1.674420e+06   1.361580e+06

In [15]:
df.isnull().sum()

video_id           0
trending_date      0
publish_time       0
channel_title      0
category_id        0
category_name      0
title              0
description      570
tags               0
views              0
likes              0
dislikes           0
comment_count      0
dtype: int64

We see that the only column with missing values if **Description**. Instead of removing these rows, I am going to make this a description of its own (either by replacing null values with 'No Description' or '-').

In [16]:
null_descrips = df.description.isnull()
null_cats = pd.Series(df.category_name[null_descrips])
null_cats.describe()

count                570
unique                14
top       People & Blogs
freq                 149
Name: category_name, dtype: object

These are the various categories in df_US that have videos where the descriptions are null. We see that the category which has the maximum number of videos with missing descriptions is 'People & Blogs.'

In [17]:
len(df.video_id.unique())

6351

We see that there are only 6351 video_ids out of 40949 rows in df_US that are unique. This could be because of a number of reasons. The video could have trended at different instances of time, say once when it was published and another when something the creator published on Twitter caused a lot of people to watch an old video of theirs. Another possible reason could be that the dataset keeps outputting a new record for every day that the video remains trending.

In [18]:
df['days_since_last_trend'] = df.groupby('video_id')['trending_date'].diff()

In [19]:
df[df['days_since_last_trend'] == '4 days']

video_id trending_date        publish_time  \
1177   XrDlDj9DfZs    2017-11-19 2017-11-14 06:08:58   
15116  DJFeaRWJdy8    2018-01-30 2018-01-25 16:00:03   
16577  qXU2qTRjBKU    2018-02-06 2018-01-31 17:32:26   
18165  8h2rlhsN9DI    2018-02-14 2018-02-05 15:00:00   
21358  rJQYzX6Bgio    2018-03-02 2018-02-17 15:11:09   
21359  XzvHSMEBFjE    2018-03-02 2018-02-17 06:23:55   
21360  rZQepOFnYi8    2018-03-02 2018-02-16 17:14:27   
21361  Mmn0zFalrD4    2018-03-02 2018-02-17 14:52:06   
21362  h_cfN1t3flE    2018-03-02 2018-02-16 23:32:15   
21363  BJDGBNFxO7o    2018-03-02 2018-02-16 17:00:01   
21364  Nl0iGMkxwcc    2018-03-02 2018-02-20 20:21:13   
21365  hN8mNXa3ICo    2018-03-02 2018-02-16 17:00:34   
21366  kl_jxVID3PE    2018-03-02 2018-02-16 18:22:38   
28510  YJ25apqV-C8    2018-04-07 2018-03-26 18:24:01   
28547  fPeQR-_JBzU    2018-04-07 2018-03-23 19:05:39   
34594  1Aoc-cd9eYs    2018-05-14 2018-05-02 16:02:35   
34604  j1v2A5pyGlY    2018-05-14 2018-05-03 04:13:34   
34607  r-luQokOiBw    2018-05-14 2018-05-02 16:00:56   
34616  efNXlVeeLIw    2018-05-14 2018-05-02 13:00:03   
34696  YKWGQyz-oLw    2018-05-14 2018-04-22 14:30:01   
34697  532MxrRNJcM    2018-05-14 2018-04-23 15:00:01   
34698  MWbQdgjsvJI    2018-05-14 2018-04-22 15:21:24   
34699  LnFsp6iM5dg    2018-05-14 2018-04-22 18:38:11   
34700  C54cHGCwBdk    2018-05-14 2018-04-23 16:00:12   
34701  9POblaMln4Q    2018-05-14 2018-04-22 21:35:35   

                                channel_title  category_id  \
1177                   Dancing With The Stars           24   
15116                          First We Feast           26   
16577                                theSkimm           22   
18165                          Inside Edition           25   
21358                      Fancy Vlogs By Gab           22   
21359                            Chris Smoove           17   
21360                              Brian Hull           24   
21361                 Hydraulic Press Channel           28   
21362                             Jackie Aina           26   
21363                              Will Smith           24   
21364                         Great Big Story           22   
21365                                   Tasty           26   
21366                                  Genius           10   
28510                         Lip Sync Battle           24   
28547                            barnabydixon            1   
34594                               Cobra Kai           24   
34604  The Tonight Show Starring Jimmy Fallon           23   
34607                             Vanity Fair           24   
34616                            TheEllenShow           24   
34696                           WatchMojo.com           24   
34697                              Tanya Burr           26   
34698                            Amber Scholl           26   
34699                                FaZe Rug           20   
34700                          USA Gymnastics           17   
34701                           MamaMiaMakeup           23   

              category_name  \
1177          Entertainment   
15116         Howto & Style   
16577        People & Blogs   
18165       News & Politics   
21358        People & Blogs   
21359                Sports   
21360         Entertainment   
21361  Science & Technology   
21362         Howto & Style   
21363         Entertainment   
21364        People & Blogs   
21365         Howto & Style   
21366                 Music   
28510         Entertainment   
28547      Film & Animation   
34594         Entertainment   
34604                Comedy   
34607         Entertainment   
34616         Entertainment   
34696         Entertainment   
34697         Howto & Style   
34698         Howto & Style   
34699                Gaming   
34700                Sports   
34701                Comedy   

                                                   title  \
1177   Lindsey and​ Mark’s - Iconic Dance - Dancing w...   
15116  Sasha 

In [20]:
df[df['video_id'] == 'XrDlDj9DfZs']

video_id trending_date        publish_time           channel_title  \
236   XrDlDj9DfZs    2017-11-15 2017-11-14 06:08:58  Dancing With The Stars   
1177  XrDlDj9DfZs    2017-11-19 2017-11-14 06:08:58  Dancing With The Stars   

      category_id  category_name  \
236            24  Entertainment   
1177           24  Entertainment   

                                                  title  \
236   Lindsey and​ Mark’s - Iconic Dance - Dancing w...   
1177  Lindsey and​ Mark’s - Iconic Dance - Dancing w...   

                                            description  \
236   Lindsey Stirling and Mark Ballas accompanied b...   
1177  Lindsey Stirling and Mark Ballas accompanied b...   

                                                   tags   views  likes  \
236   abc|"dancing"|"stars"|"dwts"|"Lindsey Stirling...   60564    975   
1177  abc|"dancing"|"stars"|"dwts"|"Lindsey Stirling...  107919   1440   

      dislikes  comment_count days_since_last_trend  
236         66            106                   NaT  
1177        90            154                4 days

We see that a video can trend on non-consecutive days, and hence just taking the first or last trending day does not contain the full picture of trending videos.

In [21]:
df.dtypes

video_id                          object
trending_date             datetime64[ns]
publish_time              datetime64[ns]
channel_title                     object
category_id                        int64
category_name                     object
title                             object
description                       object
tags                              object
views                              int64
likes                              int64
dislikes                           int64
comment_count                      int64
days_since_last_trend    timedelta64[ns]
dtype: object

Let us see how many unique channel names there are in our dataset:

In [22]:
len(df.channel_title.unique())

2207

In [23]:
len(df.title.unique())

6455

It is interesting that the humber of unique video IDs is not equal to the number of unique video titles. I will examine this further while performing EDA.

Let us see if uppercase titles are most commonly found among trending videos than lowercase/regularly capitalized ones:

In [24]:
n = 0
for charac in df.title.unique():
    if charac == charac.upper():
        n += 1
    else:
        continue
print(n)

348


We see that out of 6455 unique video titles in our dataset, 348 are in uppercase, which shows that it is in the minority.

#### Filling NaN values

In [25]:
df['description'].fillna(value='No Description', inplace=True)

In [26]:
from datetime import timedelta

In [27]:
df['days_since_last_trend'].fillna(value=timedelta(seconds=0), inplace=True)

We see that df_US has a large number of videos with web links included in their description! Maybe this is key to trend in the 21st Century? 

When I perform EDA, I am curious to examine what are the most common websites that are linked. Is it other social media belonging to the creator? Is it other Youtube links?

In [28]:
df.head(20)

video_id trending_date        publish_time          channel_title  \
0   2kyS6SvSYSE    2017-11-14 2017-11-13 17:13:01           CaseyNeistat   
1   1ZAPwfrtAFY    2017-11-14 2017-11-13 07:30:00        LastWeekTonight   
2   5qpjK5DgCt4    2017-11-14 2017-11-12 19:05:24           Rudy Mancuso   
3   puqaWrEC7tY    2017-11-14 2017-11-13 11:00:04  Good Mythical Morning   
4   d380meD0W0M    2017-11-14 2017-11-12 18:01:41               nigahiga   
5   gHZ1Qz0KiKM    2017-11-14 2017-11-13 19:07:23               iJustine   
6   39idVpFF7NQ    2017-11-14 2017-11-12 05:37:17    Saturday Night Live   
7   nc99ccSXST0    2017-11-14 2017-11-12 21:50:37     CrazyRussianHacker   
8   jr9QtXwC9vc    2017-11-14 2017-11-13 14:00:23       20th Century Fox   
9   TUmyygCMMGA    2017-11-14 2017-11-13 13:45:16                    Vox   
10  9wRQljFNDW8    2017-11-14 2017-11-13 02:05:26                    NFL   
11  VifQlJit6A0    2017-11-14 2017-11-13 03:00:00                    amc   
12  5E4ZBSInqUU    2017-11-14 2017-11-13 17:00:00             marshmello   
13  GgVmn66oK_A    2017-11-14 2017-11-12 14:00:00          NowThis World   
14  TaTleo4cOs8    2017-11-14 2017-11-12 18:30:01        The king of DIY   
15  kgaO45SyaO4    2017-11-14 2017-11-13 20:09:58         BostonDynamics   
16  ZAQs-ctOqXQ    2017-11-14 2017-11-12 17:00:05                Cracked   
17  YVfyYrEmzgM    2017-11-14 2017-11-13 16:00:07                 TED-Ed   
18  eNSN6qet1kE    2017-11-14 2017-11-13 15:30:17            PeterSripol   
19  B5HORANmzHw    2017-11-14 2017-11-12 22:00:01                SciShow   

    category_id         category_name  \
0            22        People & Blogs   
1            24         Entertainment   
2            23                Comedy   
3            24         Entertainment   
4            24         Entertainment   
5            28  Science & Technology   
6            24         Entertainment   
7            28  Science & Technology   
8             1      Film & Animation   
9            25       News & Politics   
10           17                Sports   
11           24         Entertainment   
12           10                 Music   
13           25       News & Politics   
14           15        Pets & Animals   
15           28  Science & Technology   
16           23                Comedy   
17           27             Education   
18           28  Science & Technology   
19           27             Education   

                                                title  \
0                  WE WANT TO TALK ABOUT OUR MARRIAGE   
1   The Trump Presidency: Last Week Tonight with J...   
2   Racist Superman | Rudy Mancuso, King Bach & Le...   
3                    Nickelback Lyrics: Real or Fake?   
4                            I Dare You: GOING BALD!?   
5                               2 Weeks with iPhone X   
6           Roy Moore & Jeff Sessions Cold Open - SNL   
7                 5 Ice Cream Gadgets put to the Test   
8   The Greatest Showman | Official Trailer 2 [HD]...   
9   Why the rise of the robots won’t mean the end ...   
10  Dion Lewis' 103-Yd Kick Return TD vs. Denver! ...   
11  (SPOILERS) 'Shiva Saves the Day' Talked About ...   
12         Marshmello - Blocks (Official Music Video)   
13             Which Countries Are About To Collapse?   
14                           SHOPPING FOR NEW FISH!!!   
15                                   The New SpotMini   
16   One Change That Would Make Pacific Rim a Classic   
17  How does your body know you're full? - Hilary ...   
18                         HomeMade Electric Airplane   
19           Founding An Inbreeding-Free Space Colony   

                                          description  \
0   SHANTELL'S CHANNEL - https://www.youtube.com/s...   
1   One year after the presidential election, John...   
2   WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...   
3   Today we find out if Link is a Nickelback amat...   
4   I know it's been a while since we did this sho...   
5   Usin

In [29]:
df.dtypes

video_id                          object
trending_date             datetime64[ns]
publish_time              datetime64[ns]
channel_title                     object
category_id                        int64
category_name                     object
title                             object
description                       object
tags                              object
views                              int64
likes                              int64
dislikes                           int64
comment_count                      int64
days_since_last_trend    timedelta64[ns]
dtype: object

Saving the dataframes:

In [30]:
file = open('US.pkl','wb')

pickle.dump(df, file)
file.close()